File: MCRunner/Strategy/StrategyRunner.cs
using OrderCreator = MCRunner.Orders.OrderCreator;
using PowerLanguage.Strategy;
using PowerLanguage;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Reflection;
using System;

namespace MCRunner.Strategy
{
    public class StrategyRunner<T> where T : SignalObject
    {
        public T Strategy { get; private set; }

        private readonly Type type = typeof(T);

        public StrategyRunner(
            IEnumerable<IInstrument> barsData = null,
            IOrderCreator orderCreator = null,
            IOutput output = null,
            IStrategyPerformance strategyInfo = null)
        {
            Strategy = (T)Activator.CreateInstance(type, "");

            if (barsData is null)
            {
                SetProtectedProperty("BarsData", new List<IInstrument>());
            }
            else
            {
                SetProtectedProperty("BarsData", new List<IInstrument>(barsData));
            }

            if (orderCreator is null)
            {
                orderCreator = new OrderCreator();
            }

            if (output is null)
            {
                output = new ConsoleOutput();
            }

            if (strategyInfo is null)
            {
                strategyInfo = new StrategyPerformance();
            }

            SetProtectedProperty("OrderCreator", orderCreator);
            SetProtectedProperty("Output", output);
            SetProtectedProperty("StrategyInfo", strategyInfo);
        }

        public void Create()
        {
            CallProtectedMethod("Create");
        }

        public void StartCalc()
        {
            CallProtectedMethod("StartCalc");
        }

        public void CalcBar()
        {
            CallProtectedMethod("CalcBar");
        }

        protected void CallProtectedMethod(string name)
        {
            MethodInfo method = Strategy.GetType().GetMethod(
                name, BindingFlags.NonPublic | BindingFlags.Instance);
            method.Invoke(Strategy, null);
        }

        protected void SetProtectedProperty(string name, object value)
        {
            PropertyInfo property = Strategy.GetType().GetProperty(
                name, BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
            property.SetValue(Strategy, value);
        }
    }
}

File: python/mcrunner/strategy/strategy_runner.py
from typing import Iterable
from ..instruments import Bars
from ..orders import OrderCreator
from .strategy_performance import StrategyPerformance

class StrategyRunner:
    def __init__(self, strategy_cls, bars_data: Iterable[Bars] = None,
                 order_creator: OrderCreator = None,
                 strategy_info: StrategyPerformance = None):
        self.Strategy = strategy_cls()
        self._bars = list(bars_data or [])
        self._order_creator = order_creator or OrderCreator()
        self._strategy_info = strategy_info or StrategyPerformance()
        # manager is expected to be handled by caller
        # supply strategy with required attributes
        self.Strategy.Bars = self._bars[0] if self._bars else Bars()
        self.Strategy.OrderCreator = self._order_creator
        self.Strategy.StrategyInfo = self._strategy_info

    def create(self):
        if hasattr(self.Strategy, 'create'):
            self.Strategy.create()

    def start_calc(self):
        if hasattr(self.Strategy, 'start_calc'):
            self.Strategy.start_calc()

    def calc_bar(self):
        if hasattr(self.Strategy, 'calc_bar'):
            self.Strategy.calc_bar()
